### DISCUSSION TITLE
Persist documentation for Map & Set missing storage and retrieval of version

### DISCUSSION THREAD

**Discussion by melutovich** [pmndrs/zustand#2689] [Q&A]
https://zustand.docs.pmnd.rs/integrations/persisting-store-data#how-do-i-use-it-with-map-and-set

The above linked sample code is missing the storing and retrieval of the "version", I've included it below, your thoughts? 

`  storage: {`
`    getItem: (name) => {`
`      const str = localStorage.getItem(name);`
`      if (!str) return null;`
`      const { state, version } = JSON.parse(str);`
`      return {`
`        state: {`
`          ...state,`
`          transactions: new Map(state.transactions),`
`        },`
`        version`
`      }`
`    },`
`    setItem: (name, newValue: StorageValue<BearState>) => {`
`      // functions cannot be JSON encoded`
`      const str = JSON.stringify({`
`        state: {`
`          ...newValue.state,`
`          transactions: Array.from(newValue.state.transactions.entries()),`
`        },`
`        version: newValue.version`
`        //version: newValue.version || 0  // perhaps this line instead to ensure there is some value`
`      })`
`      localStorage.setItem(name, str)`
`    },`
`    removeItem: (name) => localStorage.removeItem(name),`
`  },`

**Comment by dai-shi** [comment]
`version` is optional, so the example isn't incorrect per se?

**Reply by melutovich** [reply]
@dai-shi

But version might exist (and I'll go so far to suggest it is better it does exist per below) so making version disappear as in the example I'd say can lead to unexpected behavior.

Here is fuller senario in which I discovered this, you can decide if this is to be totally ignored, a suggestion for the example documentation, and/or a bug/improvement in the code.

As we are using Maps, I copied originally from the example

We did a breaking change adjustment to our store.  Reading though the persist documentation, I understood to change the version and create a migration.

Based on https://zustand.docs.pmnd.rs/integrations/persisting-store-data#version
I assumed incorrectly the stored version was 0 (the default) [I had not originally set the version to 0, but based on 0 being the default had assumed there was a version of 0 stored, but even if I had set the version in the options the example ignores it (without a warning)]
a)The migrate with a version of 0 was never run
b)the existing now broken stored value was kept

"a" and "b" because there was no stored version of 0 and also ("b") the zustand persist code has a typeof number check:
`            typeof deserializedStorageValue.version === 'number' &&`
`            deserializedStorageValue.version !== options.version`

this might be better similar to:
`            ((typeof deserializedStorageValue.version === 'number' &&`
`            deserializedStorageValue.version !== options.version)`
`            || (!'version' in deserializedStorageValue &&`
`            typeof options.version === 'number' &&`
`            options.version !== 0)`

If later we decide to make a breaking change adjustment to our store I'd like the ability to migrate, so I want to make sure some version exists in our store and is read back, hence my suggested update to the example.

**Reply by dai-shi** [reply]
> so making version disappear as in the example I'd say can lead to unexpected behavior.

I thought it's minor, but as you experienced, it's convincing.
Can you send a PR to fix the docs?

**Reply by melutovich** [reply]
@dai-shi 
Above, I suggested two possible lines in the "set"; I lean toward just the latter, how should I include it/them in the PR?

`        version: newValue.version`
 `       //version: newValue.version || 0  // perhaps this line instead to ensure there is some value`
 
 Further, should this PR, be by way of an issue, so that a release notes can include it so more people are aware?

**Reply by dai-shi** [reply]
I think it should be kept optional. Let's use object spread. The `version` should never become `0` implicitly.

Release notes point to the PR. But, if you prefer feel free to open an issue.

**Reply by melutovich** [reply]
https://github.com/pmndrs/zustand/pull/2692
